### PR TITLE
fix: check if update is necessary before adding sync request.

### PR DESCRIPTION
**Pull request by nikkhilmuthye**
**Issue #, if available:**

**Description of changes:**
Do not add the sync request in the queue if the update is not necessary. So we need to check that before we add it to the queue.
Also added some new booleans for tests to ensure that the queue is actually empty. 

**Why is this change necessary:**
We should not have a sync request which does not need to be processed in the queue. This was causing issue in the periodic sync wherein a local update was getting converted to a full sync due to an unnecessary local update request after the last cloud shadow update.

**How was this change tested:**
Updated the tests.

**Any additional information or context required to review the change:**

**Checklist:**
 - [ ] Updated the README if applicable
 - [X] Updated or added new unit tests
 - [X] Updated or added new integration tests
 - [ ] If your code makes a remote network call, it was tested with a proxy
 
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/integrationtests/java/com/aws/greengrass/integrationtests/SyncTest.java [hunk 1]
```diff
@@ -112,9 +116,11 @@
 
     private Supplier<Optional<SyncInformation>> syncInfo;
     private Supplier<Optional<ShadowDocument>> localShadow;
+    private final Lock sequential = new ReentrantLock();
 
     @BeforeEach
     void setup() {
+        sequential.lock();
```
Comment (rbattle) [inline src/integrationtests/java/com/aws/greengrass/integrationtests/SyncTest.java:123]: why do you need a lock in setup / cleanup?

File: src/main/java/com/aws/greengrass/shadowmanager/sync/strategy/BaseSyncStrategy.java [hunk 2]
```diff
@@ -95,6 +98,29 @@
...
+     * @return true if Shadow Manger is syncing.
+     */
+    public boolean isSyncing() {
+        return syncing.get();
+    }
+
+    /**
+     * Getter for executing boolean.
+     *
+     * @return true if Shadow Manger is executing.
```
Comment (rbattle) [inline src/main/java/com/aws/greengrass/shadowmanager/sync/strategy/BaseSyncStrategy.java:118]: nit: Manager

File: src/main/java/com/aws/greengrass/shadowmanager/sync/strategy/BaseSyncStrategy.java [hunk 3]
```diff
@@ -95,6 +98,29 @@
...
 
+    /**
+     * Indicates whether a sync request is executing or not.
+     */
+    AtomicBoolean executing = new AtomicBoolean(false);
+
+    /**
+     * Getter for syncing boolean.
+     *
+     * @return true if Shadow Manger is syncing.
```
Comment (rbattle) [inline src/main/java/com/aws/greengrass/shadowmanager/sync/strategy/BaseSyncStrategy.java:109]: nit: Manager

File: src/integrationtests/java/com/aws/greengrass/integrationtests/SyncTest.java (hunk 1 above)
Comment (nikkhilmuthye) [reply src/integrationtests/java/com/aws/greengrass/integrationtests/SyncTest.java]: This was one of the options I had tried to make the tests less flaky. I thought I had removed it. But removed it now.

**Review by rbattle — APPROVED** [review]

**Review by vabhasin — APPROVED** [review]
